PizzaOrderingSystem/src/app/effects/pizza-auth.effects.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  setLoadingSpinner,
+  setErrorMessage,
+} from '../shared/action/shared.actions';
+
+import { AuthService } from '../services/auth.service';
+import { OrderService } from '../services/order.service';
+import { exhaustMap, map, catchError, tap, mergeMap } from 'rxjs/operators';
+import { autoLogout, loginFail, loginStart, loginSuccess } from '../actions/pizza-auth.action';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { Injectable } from '@angular/core';
+import { Store } from '@ngrx/store';
+// import { AppState } from 'src/app/store/app.state';
+import { of } from 'rxjs';
+import { Router } from '@angular/router';
+import { Pizza } from '../models/Pizza';
+import { LoginData } from '../models/LoginData';
+
+@Injectable()
+export class AuthEffects {
+  constructor(
+    private actions$: Actions,
+    private authService: AuthService,
+    private orderService: OrderService,
+    private store: Store<Pizza>,
+    private router: Router
+  ) { }
+
+  //This will trigger for login operation.
+  login$ = createEffect(() => {
+    return this.actions$.pipe(
+      ofType(loginStart),
+      exhaustMap((action: any) => {
+       
+        return this.authService.login(action.userData).pipe(
+          map((data) => {
+            this.store.dispatch(setLoadingSpinner({ status: false }));
+            this.store.dispatch(setErrorMessage({ message: '' }));
```

Comment:
It's not a best practice to dispatch multiple actions from an effect. What you should do is return the success or failed action from the effect. And catch this action inside the spinner reducer.

https://twitter.com/mikeryandev/status/938108618133602304

Reply:
Removed the multiple actions from an effect, and Pushed the code to the server feature/login-page